File: go/test/legacy_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"
	"time"

	sdk "voxgigaaregurusdk"
	"voxgigaaregurusdk/core"

	vs "github.com/voxgig/struct"
)

func TestLegacyEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Legacy(nil)
		if ent == nil {
			t.Fatal("expected non-nil LegacyEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := legacyBasicSetup(nil)
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		legacyRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.legacy", setup.data)))
		var legacyRef01Data map[string]any
		if len(legacyRef01DataRaw) > 0 {
			legacyRef01Data = core.ToMapAny(legacyRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = legacyRef01Data

		// LOAD
		legacyRef01Ent := client.Legacy(nil)
		legacyRef01MatchDt0 := map[string]any{
			"id": legacyRef01Data["id"],
		}
		legacyRef01DataDt0Loaded, err := legacyRef01Ent.Load(legacyRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		legacyRef01DataDt0LoadResult := core.ToMapAny(legacyRef01DataDt0Loaded)
		if legacyRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if legacyRef01DataDt0LoadResult["id"] != legacyRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func legacyBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "legacy", "LegacyTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read legacy test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse legacy test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"legacy01", "legacy02", "legacy03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	env := envOverride(map[string]any{
		"AAREGURU_TEST_LEGACY_ENTID": idmap,
		"AAREGURU_TEST_LIVE":      "FALSE",
		"AAREGURU_TEST_EXPLAIN":   "FALSE",
		"AAREGURU_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["AAREGURU_TEST_LEGACY_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["AAREGURU_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["AAREGURU_APIKEY"],
			},
			extra,
		})
		client = sdk.NewAareguruSDK(core.ToMapAny(mergedOpts))
	}

	return &entityTestSetup{
		client:  client,
		data:    entityData,
		idmap:   idmapResolved,
		env:     env,
		explain: env["AAREGURU_TEST_EXPLAIN"] == "TRUE",
		now:     time.Now().UnixMilli(),
	}
}

File: rb/test/legacy_entity_test.rb
# Legacy entity test

require "minitest/autorun"
require "json"
require_relative "../Aareguru_sdk"
require_relative "runner"

class LegacyEntityTest < Minitest::Test
  def test_create_instance
    testsdk = AareguruSDK.test(nil, nil)
    ent = testsdk.Legacy(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = legacy_basic_setup(nil)
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    legacy_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.legacy")))
    legacy_ref01_data = nil
    if legacy_ref01_data_raw.length > 0
      legacy_ref01_data = Helpers.to_map(legacy_ref01_data_raw[0][1])
    end

    # LOAD
    legacy_ref01_ent = client.Legacy(nil)
    legacy_ref01_match_dt0 = {
      "id" => legacy_ref01_data["id"],
    }
    legacy_ref01_data_dt0_loaded, err = legacy_ref01_ent.load(legacy_ref01_match_dt0, nil)
    assert_nil err
    legacy_ref01_data_dt0_load_result = Helpers.to_map(legacy_ref01_data_dt0_loaded)
    assert !legacy_ref01_data_dt0_load_result.nil?
    assert_equal legacy_ref01_data_dt0_load_result["id"], legacy_ref01_data["id"]

  end
end

def legacy_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "legacy", "LegacyTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = AareguruSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["legacy01", "legacy02", "legacy03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  env = Runner.env_override({
    "AAREGURU_TEST_LEGACY_ENTID" => idmap,
    "AAREGURU_TEST_LIVE" => "FALSE",
    "AAREGURU_TEST_EXPLAIN" => "FALSE",
    "AAREGURU_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["AAREGURU_TEST_LEGACY_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["AAREGURU_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["AAREGURU_APIKEY"],
      },
      extra || {},
    ])
    client = AareguruSDK.new(Helpers.to_map(merged_opts))
  end

  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["AAREGURU_TEST_EXPLAIN"] == "TRUE",
    now: (Time.now.to_f * 1000).to_i,
  }
end
